### DISCUSSION TITLE
Proposal: Single-line if statements

### DISCUSSION THREAD

**Discussion by MisinformedDNA** [dotnet/csharplang#1785] [Language Ideas]
It's frequent to have an `if` statement that contains a single statement. Due to design guidelines, we often end up wrapping those statements in braces. It would be great if we could find a way to remove the braces and still satisfy common guidelines. 

Previous proposals I've seen have tried to make this extend to multi-block statements like if-else or try-catch, but those have proven to have issues. Instead, this proposal is limited to single-line `if` statements that _can_ be written without braces and also **do not** have an `else` statement.

So the scope for this proposal is code that can be written as

```csharp
if (condition) statement;
```

but usually ends up formatted as

```csharp
if (condition)
{
    statement;
}
```

Here's some examples to consider:

```csharp
// 1.
if (conditional)
{
    throw new ArgumentNullException(nameof(arg));
}

// 2.
if (conditional)
{
    _instance = new SomeSingleton();
}

// 3.
if (conditional)
{
    return false;
}

// 4.
if (conditional)
{
    return;
}
```

**Option A: Arrow expressions:** `if (condition) => expression`

```csharp
// 1.
if (arg == null) => throw new ArgumentNullException(nameof(arg));

// 2.
if (_instance == null) => _instance = new SomeSingleton();

// 3.
if (conditional) => return false;    // ISSUE: Does not behave like other arrow expressions

// 4.
if (conditional) => return;    // ISSUE: Does not behave like other arrow expressions
```

**Option B: Arrow expressions with conditional jumps/returns:** `[jump] if (conditional) [=> expression]`

```csharp
// 1.
if (arg == null) => throw new ArgumentNullException(nameof(arg));

// 2.
if (_instance == null) => _instance = new SomeSingleton();

// 3.
return if (conditional) => false;    // Seems odd that return and false are separated

// 4.
return if (conditional);

// 5. Also would allow for other jump statements like
continue if (conditional);
break if (conditional);
```

**Option C: Reorder the syntax:** `statement if (conditional)`

```csharp
// 1. 
throw new ArgumentNullException(nameof(arg)) if (conditional);

// 2. 
_instance = new SomeSingleton() if (conditional);

// 3. 
return false if (conditional);

// 4. 
return if (conditional);
```

Option C may seem odd at first since the order is backwards from how we are used to, but the proposed C# 8 `switch expression` also switched the `switch` keyword and the match expression, so there is some precedence for this.

**Option D: Reorder the syntax, but use `when`:** `statement when conditional`
Given the similarities between Options C and other uses of the [`when` keyword](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/when), a new option becomes intriguing:

```csharp
// 1. 
throw new ArgumentNullException(nameof(arg)) when conditional;

// 2. 
_instance = new SomeSingleton() when conditional;

// 3. 
return false when conditional;

// 4. 
return when conditional;
```

In conclusion, I think option D covers a simple but very common use case. Allowing us to condense our code from 

```csharp
if (condition)
{
    statement;
}
```

to

```csharp
statement when condition;
```

Someone will want to expand this to cover `else` blocks, but I don't see that as stictly necessary. In fact, if someone _really_ wanted an `else` block, and all other options didn't meet their needs, they could easily use the rule above to create a faux `else` just by doing:

```csharp
return 0 when conditional;
value = 1 when !conditional;
```

Thoughts?

**Comment by HaloFour** [comment]
`if (condition) statement;` is already perfectly legal and quite succinct.

**Comment by HaloFour** [comment]
Also #1493, #724, #453, #325 and probably a few others

**Comment by jnm2** [comment]
Allowing a superfluous `=>` before embedded statements is covered by https://github.com/dotnet/csharplang/issues/1210 and similar to https://github.com/dotnet/csharplang/issues/1540 and https://github.com/dotnet/csharplang/issues/1567.

**Comment by masonwheeler** [comment]
This is unlikely to ever gain acceptance, but if you like the idea, have a look at the Boo language, where option C is supported and considered idiomatic. 😄

**Comment by MisinformedDNA** [comment]
> `if (condition) statement;` is already perfectly legal and quite succinct.

@HaloFour I see that you like to bring this up on various proposals, but there's a huge difference that you repeatedly miss.

`if (condition) statement;` is legal _but discouraged_. We all know why it's discouraged and we all know that many enterprises don't allow this already discouraged code,, the .NET team being one of them (see [example](https://github.com/dotnet/corefx/blob/e0ba7aa8026280ee3571179cc06431baf1dfaaac/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/TypeCode.cs)).

The goal of this proposal (and others like it) is to find a way to write the code in a single line that is legal, succinct, readable, and not ambiguous (and thus not discouraged). This follows the same line of thought that brought about expression-bodied members.

And yes, it has been brought up multiple times, because multiple people would like to find a solution. This issue would encompass #1493, #724. #453 and #325 change the syntax a bit, but are after the same goal.

**Comment by jnm2** [comment]
if `if (arg == null) => throw new ArgumentNullException(nameof(arg));` is good, why isn't `if (arg == null) throw new ArgumentNullException(nameof(arg));` even better? And if your style guide would allow a less good (`=>`) syntax but not the better syntax, shouldn't the style guide be changed?

**Comment by MisinformedDNA** [comment]
> Allowing a superfluous => before embedded statements is covered by #1210 and similar to #1540 and #1567

@jnm2 The problem with #1210 and #1567 is that they are trying to apply a construct that works well for single- line, single-block statements and try to apply them to multi-block cases, which hasn't worked out great. This proposal avoids that. #1540 wanted expression-bodied members to accept statement-bodied members, which I reject in option A.

**Comment by HaloFour** [comment]
@MisinformedDNA 

There's nothing ambiguous about `if (condition) statement;`.  There's something ambiguous about `if (condition) statement; statement;` and that same ambiguity also exists in `if (condition) => statement; statement`;

I don't like the idea of inventing new syntax just to workaround overzealous code style requirements.  When those same enterprises then forbid this syntax on the same grounds are you going to suggest another syntax to avoid those requirements?

**Comment by MisinformedDNA** [comment]
> but if you like the idea, have a look at the Boo language, where option C is supported and considered idiomatic

@masonwheeler No need to be facetious. Obviously, I'm here because I like C#. I'm not going to change languages over any single feature!

That said, this construct is also fully supported by Perl and Ruby, so there are at least three languages that already support this.

**Comment by CyrusNajmabadi** [comment]
> Option A:
```c#
// 1.
if (arg == null) => throw new ArgumentNullException(nameof(arg));
```

Why not just do:

```c#
if (arg == null) { throw new ArgumentNullException(nameof(arg)); }
```

It's only one character longer...

**Comment by scalablecory** [comment]
Your proposal does not enable something that was previously impossible, make easy something that was previously hard, increase readability, or increase productivity.

That's a no for me.

**Comment by jnm2** [comment]
@MisinformedDNA 

> @masonwheeler No need to be facetious. Obviously, I'm here because I like C#. I'm not going to change languages over any single feature!

I don't think he meant that facetiously. He's one of the Boo language maintainers. I'd point out something cool about my language too if I was him. :D

**Comment by MisinformedDNA** [comment]
> if (arg == null) => throw new ArgumentNullException(nameof(arg)); is good, why isn't if (arg == null) throw new ArgumentNullException(nameof(arg)); even better? And if your style guide would allow a less good (=>) syntax but not the better syntax, shouldn't the style guide be changed?

@jnm2 `if (arg == null) throw new ArgumentNullException(nameof(arg));` may be fewer characters, but it's still ambiguous.

@jnm2 @CyrusNajmabadi
`if (arg == null) => throw new ArgumentNullException(nameof(arg));` is not the best example because if we implemented arrow expressions like other places, it would work fine. The following examples are more useful because they _don't_ work like the others.

```csharp
// 3.
if (conditional)
{
    return false;
}

// 4.
if (conditional)
{
    return;
}
```

Option A would lead to:

```csharp
// 3.
if (conditional) => return false;    // ISSUE: Does not behave like other arrow expressions.

// 4.
if (conditional) => return;    // ISSUE: Does not behave like other arrow expressions
```

and Option B would lead to 

```csharp
// 3.
return if (conditional) => false;    // Seems odd that return and false are separated

// 4.
return if (conditional);
```

Option A causes a problems with a pre-existing construct and Option B has weird code separation.

**Comment by jnm2** [comment]
@MisinformedDNA 

> > if `if (arg == null) => throw new ArgumentNullException(nameof(arg));` is good, why isn't `if (arg == null) throw new ArgumentNullException(nameof(arg));` even better? And if your style guide would allow a less good (`=>`) syntax but not the better syntax, shouldn't the style guide be changed?
> 
> if (arg == null) throw new ArgumentNullException(nameof(arg)); may be fewer characters, but it's still ambiguous.

It's not ambiguous either to the compiler or to anyone I have seen code, so my question still stands.

❔

**Comment by CyrusNajmabadi** [comment]
> @jnm2 @CyrusNajmabadi
if (arg == null) => throw new ArgumentNullException(nameof(arg)); is not the best example because if we implemented arrow expressions like other places, it would work fine. The following examples are more useful because they don't work like the others.

Why not just write:

> // 3.
```c#
if (conditional) { return false; }
if (conditional) { return; }
```

You can already write a single-line-if that is unambiguous and easy to understand.  It's not clear to why any new syntax would be required here.  At best, the proposed syntaxes save 1-2 chars... that seems so completely unnecessary to support.

**Comment by theunrepentantgeek** [comment]
Personally, I quite like _Option C_; I've seen this in other languages, allowing predicates using suffixed **if** and **unless** keywords:

```
throw new InvalidOperationException() unless (value != null);
return false if (string.IsNullOrBlank(value));
```

_**BUT**_: These have been well debated previously, aren't well liked, and don't really add new power to the language. I believe they're very unlikely to happen.

**Comment by AustinBryan** [comment]
You're basically asking for a language change because you care so heavily about what other people think and violating their guidelines? What happens if `if (condition) => statement;` is added, but then it's later removed from most coding guidelines? Are you gonna request for *new* syntax for `if (condition) => => statement;` or `if (condition) -> statement;`?

Just write the code you want, you don't need their approval or the code to be changed. I myself use the singular brace whenever possible, putting the condition and statement if it can fit, if not, on separate lines.

**Comment by MisinformedDNA** [comment]
> I don't think he meant that facetiously. He's one of the Boo language maintainers. I'd point out something cool about my language too if I was him. :D

@jnm2 OK. I didn't know that. Not sure why I would've known that.

**Comment by MisinformedDNA** [comment]
@jnm2 

> if if (arg == null) => throw new ArgumentNullException(nameof(arg)); is good, why isn't if (arg == null) throw new ArgumentNullException(nameof(arg)); even better? And if your style guide would allow a less good (=>) syntax but not the better syntax, shouldn't the style guide be changed?

I don't understand the question. What you claim is better, is actually worse.
 
> > if (arg == null) throw new ArgumentNullException(nameof(arg)); may be fewer characters, but it's still ambiguous.
> 
> It's not ambiguous either to the compiler or to anyone I have seen code, so my question still stands.

It is definitely not ambiguous to the compiler, but it certainly could be ambiguous of what the original author _meant_. Like "is that line supposed to be inside the `if`? Or was it meant to be executed no matter what?"

Feel free to [read up on the topic](https://www.bing.com/search?q=why+you+should+avoid+if+statements+without+braces).

**Comment by svick** [comment]
@MisinformedDNA 

> Feel free to read up on the topic.

All the results on the first page of those results for me are about the dangers of:

```c#
if (condition)
    statement;
```

but not about:

```c#
if (condition) statement;
```

I don't understand how is the above one-line-if ambiguous. Do you have any source explaining why that form is bad?

**Comment by HaloFour** [comment]
> I don't understand how is the above one-line-if ambiguous. Do you have any source explaining why that form is bad?

Furthermore, why it's any worse than `if (condition) => statement;`, which would, quite literally, accomplish the exact same thing and suffer the exact same potential problems:

```cs
if (condition) =>
    statement;
    statement; // oops!
```

**Comment by jnm2** [comment]
> > I don't think he meant that facetiously. He's one of the Boo language maintainers. I'd point out something cool about my language too if I was him. :D
>
> OK. I didn't know that. Not sure why I would've known that.

Not sure why you'd have to know something like that in order to assume that it was a well-intentioned comment.

Also, what @svick and @HaloFour said.

In practice, automatic indentation fixes the situation as it's being typed. An author would typically have to fight the editor in order to create misunderstandable syntax. `if (...) throw ...` and `if (...)↵    throw` are so prevalent that I don't buy the argument that folks are actually making mistakes around this anymore—published opinions to the contrary notwithstanding.

**Comment by MisinformedDNA** [comment]
I feel like some of you are arguing just to argue.

@svick I know you are smart enough to know that 

```csharp
if (condition)
    statement;
```

and

```csharp
if (condition) statement;
```

are syntactically the same. Would you really treat these differently in your style guide? All it takes is one person to apply their preference of the former form and we are right back to where we started.

**Comment by HaloFour** [comment]
@MisinformedDNA 

So are:

```cs
if (condition) => statement;
```

and

```cs
if (condition) =>
    statement;
```

**Comment by MisinformedDNA** [comment]
@jnm2 
> I don't buy the argument that folks are actually making mistakes around this anymore

Apple made the mistake [only 4 years ago](https://www.imperialviolet.org/2014/02/22/applebug.html). I bet they had auto-formatting too. Even in VS, at times, you need to manually format the document.

Also, the .NET team still enforces the use of braces. See [example](https://github.com/dotnet/corefx/blob/e0ba7aa8026280ee3571179cc06431baf1dfaaac/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/TypeCode.cs). A one-line if statement would allow them to remove almost 75% of the code there.

@scalablecory As a single line, you can argue it's not more readable, but when you take a step back and consider how much more code you can see on your screen, it is certainly more.

**Comment by MisinformedDNA** [comment]
@HaloFour They are not syntactically the same, they are functionally the same. But `if (condition) statement;` leaves you open to the common bugs specified earlier. `if (condition) => statement` makes it explicit that you are executing one and only one statement.

**Comment by yaakov-h** [comment]
How does it make it explicit? How can you enforce it?

You can just as easily make the same mistakes with this:

```csharp
if (condition) =>
    statement; // expected
    statement; // oops!
```

**Comment by HaloFour** [comment]
@MisinformedDNA 

> They are not syntactically the same, they are functionally the same. 

Precisely.  Duplicate syntax for no reason at all.

> `if (condition) => statement` makes it explicit that you are executing one and only one statement.

It does nothing of the short because it doesn't (and can't) change the behavior of the compiler to actually enforce anything.

You'd be better served by an analyzer which could evaluate the whitespace around the statement.

**Comment by MisinformedDNA** [comment]
I trust the compiler, I don't trust humans.

And I don't know why everyone keeps focusing on `if (condition) => statement`. I recommended we go down the path of `statement if conditional` (or `statement when conditional`). This was after I detailed my thinking and how I got to my conclusion, which as it turns out is (almost) the exact same syntax that _other languages already use_. I'm not suggesting anything that isn't already done in other languages.

I am, apparently, unable to convince you, so I will simply challenge you to ask yourself why other language designers (of extremely popular languages) decided this was a good feature and why it specifically doesn't belong or make sense in C#.

**Comment by HaloFour** [comment]
C# has perfectly good syntax for this already.  There doesn't need to be a reason to not adopt a feature.

**Comment by CyrusNajmabadi** [comment]
@MisinformedDNA You didn't address my point above.  Why not just use curlies?  You get the best of all worlds and solve all the problems you seem to be concerned with.

**Comment by MisinformedDNA** [comment]
@CyrusNajmabadi A) It's ugly B) It doesn't follow C# style guidelines.

The goal is to make it more readable, not less.

**Comment by MisinformedDNA** [comment]
@HaloFour 
> There doesn't need to be a reason to not adopt a feature.

That's exactly what you've been providing this whole time! lol That's exactly what you provided in the previous statement.

> C# has perfectly good syntax for this already

Also, C# had a perfectly good syntax before expression-bodied members were introduced. Turns out people like the new way. A new way isn't necessarily wrong.

**Comment by bondsbw** [comment]
@MisinformedDNA

> Turns out people like the new way. A new way isn't necessarily wrong.

I don't think the general opinion is that your suggestion is wrong.  It's just that it adds little to no real value for most developers.

But it does add work in implementing the language feature (tremendous amounts actually), which can be better utilized by bringing popular and valuable features to C#.

**Comment by CyrusNajmabadi** [comment]
> A) It's ugly 

How so?  

> B) It doesn't follow C# style guidelines.

Then change the style guideline.  It seems patently ridiculous to say: we need a new language feature because the existing one that serves this purpose perfectly fine is blocked by someone's style guideline. :)

--

This thread started with:

>It's frequent to have an if statement that contains a single statement. Due to design guidelines, we often end up wrapping those statements in braces. 

So you have a design guideline that requires wrapping in braces.  But you also have a guideline that says you can't put this  on multiple lines.  And your solution to this self-imposed choice is to go and try to get C# to change to introduce a feature because you won't change your style rules?  :)

**Comment by CyrusNajmabadi** [comment]
> Also, C# had a perfectly good syntax before expression-bodied members were introduced.

This is not true.  The existing systaxes were *majorly* cumbersome.  For example, you had to write, at a minimum:

```c#
int Prop { get { return whatever; } }
```

Instead of:

```c#
int Prop => whatever;
```

In this case you saved 16 chars.  That was more than 40% the size of the declaration.  In this case there was a substantive improvement on an exceptionally common pattern seen everywhere.  It also got rid of 5 unnecessary tokens in a declaration that itself was 10 tokens to begin with.  So that halved the token count.  It also let the most important part of the declaration shine through.  Cutting out so much of hte noise and letting through all the important signal .

This is not the case with the issue being discussed here.  Several of the suggestions save on a *single* character over a perfectly fine pattern.  Indeed, the pattern solves all the listed problems.  But the reason you don't like it is because "it's ugly" :)  

Introducing an entire new syntax into the language, and then having to go update the entire tooling stack, juts because you think `=>` is prettier than `{ }` is a *really* poor use of resources :)

**Comment by AustinBryan** [comment]
@MisinformedDNA  Well, if you really want to show your intent, you could always just do this: 

```csharp
if (condition) /*=>*/ statement; 
```

**Comment by CyrusNajmabadi** [comment]
> The goal is to make it more readable, not less.

I don't see anything about `=>` that makes it more readable than `{ }` when used with an if-statement:

```c#
if (condition) { statement; }
```

Is just as readable as

```c#
if (condition) => statement;
```

Furthermore, i would say that other forms are less readable.  For example:

```c#
return if (conditional) => false;
```

Definitely less readable.  'returns' used to be unconditional, now i have to realize that i can flow past this depending on the conditional.

Or

```c#
return false if (conditional);
```

Also less readable.  I have to know the condition first to decide what to do.  Making the condition then come after the flow just makes things harder to read IMO.

This is similar to why linq declared variables up front, instead of doing it like SQL does.  There's a certain flow of declarations and order of operations that C# tends to prefer that this goes very much against.

**Comment by AustinBryan** [comment]
@CyrusNajmabadi There has been a lot of suggestions that gain only 1 character, but this one actually *looses* 3! The current syntax is even shorter than the suggested 😆

Also, 

```csharp
if (condtion) statement;

if (condition)
    statement;
```

Are perfectly clear in intent. Adding `=>` would actually make it more confusing to many more people, as many more people would assume there is a valuing, important difference between the two syntaxes, otherwise it wouldn't've been added. Then they'd have to stop what they were doing just to find out that the only thing it serves is to get out of violating style guides, which again, what if `=>` is banned from style guides in the likely event it causes confusion?

**Comment by CyrusNajmabadi** [comment]
I'm sympathetic to the issue that arises with:

```c#
if (condition)
    statement;
```

That said. I think there are perfectly good language features *today* that address it.  No clue why someone people would have a style guideline that would prevent you from using that language feature.

**Comment by theunrepentantgeek** [comment]
Way, _way_, **way** back in the deep dark mists of time, I worked on a team that was adopting C# and .NET 1.1, moving from a combination of Visual Basic 6, VBA, PowerBuilder, and Delphi. (Possibly others.)

At that time, and for that specific team, the members had the classic worry that 
```
if (condition)
    statement;
```
would be modified to 
```
if (condition)
    statement;
    statement;
```
And that the code would therefore not work as it should.

The team adopted a rule that explicit blocks should always be used. This was primarily driven by the lack of experience within the team, and it made sense for _that team_ at _that time_.

With improvements in IDEs (especially automatic code formatting), the advent of tools like R# and other analyzers (that will highlight dodgy code), and the widespread adoption of unit testing (which would typically catch issues like this), I think there's a case to be made for _revisiting the guidelines_.

Coding standards are supposed to deliver value by guiding us towards known good practices. They're supposed to work for us, not the other way around.

**Comment by AustinBryan** [comment]
With the new version VS it's even easier now because all I have to do is hit `ctrl + .` and I can add or remove the braces around so it's super quick and I never get tripped up, so I think it should be reconsidered.

**Comment by alrz** [comment]
> Due to design guidelines, we often end up wrapping those statements in braces. 

**DO** wrap `if` embedded statement in braces

**DO NOT** use `if (...) =>` syntax.

oh no.

**Comment by DavidArno** [comment]
Or

**DO** wrap if embedded statement in braces

**DO NOT** use `doSomething if (…)` syntax.

oh no again.

Design guidelines really should be enforced with analysers these days. And analysers can detect things like:
```cs
if (condition) statement1; statement2;
```
and flag it as a warning. This is always better than guidelines that need manual code reviews to enforce.

**Comment by svick** [comment]
@MisinformedDNA 

> Would you really treat these differently in your style guide?

I don't think what I would put in my style guide is relevant to this discussion. But yes, I would treat them differently, but the other way around: Allow the variant with newline, disallow the one without newline, purely based on my personal stylistic preferences.

> All it takes is one person to apply their preference of the former form and we are right back to where we started.

If I'm using ReSharper with some specific settings, then applying my preferences removes braces from `if`s where it can. So by this logic, even using braces is still dangerous. Otherwise, it seems to me like you're complaining about the behavior of some IDE (presumably VS). In that case, maybe it would make sense to change the IDE, instead of changing the language (which is much more work).

> I recommended we go down the path of `statement if conditional` (or `statement when conditional`).

I think that's even worse. `if (condition) => statement;` is just redundant, `statement if (condition);` is actively harmful.

Consider this code:

```c#
throw new Exception("Some fairly long error message explaining what the exception is about.", ex) if (condition);
```

That condition is very important, but it's hidden at the far end of the line. This makes it easy to overlook, especially since it might require scrolling to see. In my opinion, syntax that hides the most important part of a statement is not good syntax.

**Comment by jnm2** [comment]
To @HaloFour's and @DavidArno's point about analyzers, we actually https://github.com/DotNetAnalyzers/StyleCopAnalyzers with half the rules turned on. We set it up to flag incorrect indentation as a compiler error. So even beyond having to actively fight the IDE in order to make the syntax look ambiguous to humans, we would also instantly break the build.

**Comment by AustinBryan** [comment]
@jnm2 Makes a good point about IDE's. Let's face it, if you're using C#, you're probably using Visual Studio, if not, some other intelligent modern IDE that has automatic indenting that will make you be constantly going out of your way to indent the code in a way that looks confusing.

**Comment by jrmoreno1** [comment]
I’d suggest followings @CyrusNajmabadi suggestion to just use the braces on the same line, because that is what I actually do.  IMO if it’s short enough to be on the same line, it’s short enough that surrounding it with a couple of extra characters isn’t a problem. If adding a couple of braces makes the line too long, it is already too long to be a single line.

Update the style guide, if the style guide is tooling, update the tooling.

**Comment by danielbls** [comment]
I'd rather be able to use something like

someBool ? return x() : return y();

**Comment by theunrepentantgeek** [comment]
Err, @danielbls why not just write
```
return someBool ? x() : y();
```
which is already legal syntax?

**Comment by DaveSkender** [comment]
I think _Option C.4_ as a limited case shorthand is nice.  Only because there's not complex return types or expressions that might be confusing.

```csharp
return if (somecondition) isMet;
Console.WriteLine($"Condition is {isMet}");  // "Condition is false", tho for `return` carrying var is moot
```
```csharp
continue if (somecondition) isMet;
```
or perhaps only in `bool` literal cases
```csharp
bool isExitCondition = true;
return if isExitCondition;
```
```csharp
continue if isExitCondition();
```

**Comment by AndrejZigmund** [comment]
Would be extremely useful to popularize early returns/breaks.

Approach of `if (expr) return smth;` was criticized 1000 of times already (biggest for me, is that you cannot see return point at once in code, so teams are forced to overclutter code with braces)

The syntactic sugar should be very simple to implement.

Should be added to `return, yield return, continue, break, throw`.

**Reply by HaloFour** [reply]
> is that you cannot see return point at once in code

Putting the condition at the end has the problem that you can't see the condition at once, either.  A single-line `if`+`return` is already very idiomatic, and if it's too long that it's difficult to read then you already shouldn't be trying to cram it onto a single line.

**Reply by CyrusNajmabadi** [reply]
> Approach of `if (expr) return smth;` was criticized 1000 of times already (biggest for me, is that you cannot see return point at once in code, so teams are forced to overclutter code with braces)


You can't see the `if` either if it that long. In which case you should be having things on multiple lines for clarity anyways. 

Additionally, we have the idiomatic way to do this today.  This makes things more difficult as now every `return` must be examined to see if it is this form. Whereas *today* it is known that all `if`s are - conditional and all `return`s are not.
